3 new blogs (#81)
kgateway-dev/kgateway.dev: content/blog/five-learnings-from-seven-years-of-building-gloo-and-kgateway.md

---
title: Five Learnings from Seven Years of Building Gloo and kgateway
toc: false
publishDate: 2025-03-04T00:00:00-00:00
author: Lin Sun, & Yuval Kohavi
---

> We started Gloo Open Source in 2018, reaching GA in early 2019. Since then, we have grown Gloo adoption to hundreds of paid customers and numerous open-source users. We truly believe Gloo Gateway is the most mature and widely deployed Envoy-based gateway available in the market. This belief led us to [donate Gloo Gateway to the CNCF](https://www.youtube.com/watch?v=psZi_T1np4U) as ‘kgateway’ at KubeCon NA 2024, further enriching the cloud native API gateway ecosystem. In this blog, we share five key learnings from the past seven years of building [Gloo](https://github.com/solo-io/gloo)/[kgateway](https://github.com/kgateway-dev/kgateway).

Reviewer: ```suggestion
We started Gloo Open Source in 2018, reaching GA in early 2019. Since then, we have grown Gloo adoption to hundreds of paid customers and numerous open-source users. We truly believe Gloo Gateway is the most mature and widely deployed Envoy-based gateway available in the market. This belief led us to [donate Gloo Gateway to the CNCF](https://www.youtube.com/watch?v=psZi_T1np4U) as **kgateway** at KubeCon NA 2024, further enriching the cloud native API gateway ecosystem. In this blog, we share five key learnings from the past seven years of building [Gloo](https://github.com/solo-io/gloo)/[kgateway](https://github.com/kgateway-dev/kgateway).
```

When we designed Gloo, we structured it with two types of pods: Gloo and Gateway. The Gloo pod served as the control plane, while the Gateway pod ran as a programmed Envoy proxy based on user’s declarative intention via Gloo custom resources. To support multiple APIs, we introduced an intermediate Proxy custom resource (CR), which was responsible for translating various user’s declarative intentions into Envoy’s xDS representation to program the Gateway pod.

{{< reuse-image src="blog/five-learnings-1.png" width="750px" >}}

While this worked well for small to medium-scale users, it introduced challenges at larger scales. The Proxy resource grew too large for etcd, leading to performance issues in serialization and deserialization. Our critical mistake was making the Proxy resource a user-facing API rather than keeping it strictly internal. This forced us to maintain backward compatibility indefinitely, adding complexity over time. For example, there are now 4–5 ways to add listeners to the Proxy resource. To prevent this issue in kgateway, we decided to keep the Proxy resource internal.

## Rethinking Control Plane Extensibility

> When we started Gloo in 2018, Kubernetes had not yet won the container orchestration war, so we built an abstraction layer to support multiple environments like Nomad, Consul, and file-based configurations. While this seemed flexible, it introduced inefficiencies. Our translation model used a snapshot-based structure, where upstreams, routes, and listeners were processed as a single large list. This simplified translation logic but severely impacted performance which saw any small change requiring recalculating the entire snapshot.

Reviewer: ```suggestion
When we started Gloo in 2018, Kubernetes had not yet won the container orchestration war, so we built an abstraction layer to support multiple environments like Nomad, Consul, and file-based configurations. While this seemed flexible, it introduced inefficiencies. Our translation model used a snapshot-based structure, where upstreams, routes, and listeners were processed as a single large list. This simplified translation logic but severely impacted performance—any small change requiring recalculating the entire snapshot.
```

We learned that anything requiring transactional consistency must be in the same CRD. In kgateway, we provide CRDs that allow users to extend routes via extensionRef to custom resources. Since we own these CRDs, we can maintain compatibility and resolve issues effectively.

### Label Selectors

Label selection proved another scaling challenge as we can’t index them. You may be wondering why this is not a problem for kubernetes? Kubernetes calculates label selections sporadically, but Gloo resolved all label selections simultaneously, leading to an O(N^2) scaling problem. In most recent Gloo, we avoided generic label selectors in hot paths like route delegation. Instead, we predefine label-based route delegation, enabling indexing for optimized label selection calculations.

## Leveraging Istio’s Control Plane

Four years ago, we explored using Istio’s control plane to manage Gloo’s gateway proxy. Given Istio’s adoption by major enterprises like eBay and Airbnb, we expected it to handle large-scale environments effectively. Unifying control planes for mesh and gateway seemed like a natural simplification.

> {{< reuse-image src="blog/five-learnings-2.png" width="750px" >}}

Reviewer: ```suggestion
{{< reuse-image src="blog/five-learnings-2.png" width="400px" >}}
```

> {{< reuse-image src="blog/five-learnings-2.png" width="400px" >}}

Reviewer: ```suggestion
{{< reuse-image src="blog/five-learnings-2.png" width="500px" >}}
```

{{< reuse-image src="blog/five-learnings-3.png" width="750px" >}}

## Control Plane Scalability

We initially used Kubernetes [controller-runtime](https://github.com/kubernetes-sigs/controller-runtime) for dependency tracking but later discovered [krt](https://github.com/istio/istio/blob/master/pkg/kube/krt/README.md), which provides a framework for building declarative controllers. Krt solved exactly the problems we faced, eliminating the need for manual dependency tracking, indexing, and reconciliation logic.

When implementing the Kubernetes Gateway API in kgateway, we deal with numerous objects interacting through targetRefs and extensionRefs. Managing these references manually with controller-runtime proved cumbersome—it provides the capability but requires manual wiring of dependencies. This means manually creating indexes, queueing events, and triggering reconciliation. With a large number of dependencies, we encountered frequent issues related to dependency tracking and resolution. Krt solved these problems by handling dependency tracking automatically. Whenever a change occurs, only the relevant objects are retranslated instead of recalculating everything.

> Previously, Gloo’s snapshot-based model processed routes, upstreams, and other resources in a single translation cycle, meaning any small change triggered a full recalculation. Krt improves scalability by separating Proxy, Upstream, and Pod translations, ensuring that only relevant components are reprocessed when changes occur. For example, when a Pod endpoint changes, it won’t trigger the translation of the Proxy. Recognizing these benefits, we built kgateway to be fully krt-based, leveraging its efficient dependency tracking and reconciliation.

Reviewer: ```suggestion
Previously, Gloo’s snapshot-based model processed routes, upstreams, and other resources in a single translation cycle, meaning any small change triggered a full recalculation. Krt improves scalability by separating Proxy, Service/Backend, and Pod translations, ensuring that only relevant components are reprocessed when changes occur. For example, when a Pod endpoint changes, it won’t trigger the translation of the Proxy. Recognizing these benefits, we built kgateway to be fully krt-based, leveraging its efficient dependency tracking and reconciliation.
```

{{< reuse-image src="blog/five-learnings-4.png" width="750px" >}}

## Join the kgateway Community

> If any of these learnings resonate with you, join us in the [kgateway slack channel](https://kgateway.dev/slack/) and follow us on [LinkedIn](https://www.linkedin.com/company/kgateway/). We’ve also got some create free technical content available for the community to learn more about kgateway including [free hands-on labs](https://kgateway.dev/resources/labs/), [videos](https://kgateway.dev/resources/videos/) and [blogs](https://kgateway.dev/blog/). If you’re keen to help grow kgateway you can also join us at [GitHub](https://github.com/kgateway-dev/kgateway).

Reviewer: ```suggestion
If any of these learnings resonate with you, join us in the [kgateway slack channel](https://kgateway.dev/slack/) and follow us on [LinkedIn](https://www.linkedin.com/company/kgateway/). We’ve also got some free technical content available for the community to learn more about kgateway including [free hands-on labs](https://kgateway.dev/resources/labs/), [videos](https://kgateway.dev/resources/videos/) and [blogs](https://kgateway.dev/blog/). If you’re keen to help grow kgateway you can also join us at [GitHub](https://github.com/kgateway-dev/kgateway).
```
